### PR TITLE
Ad/xamarin draw

### PR DESCRIPTION
**Pull request by AndyDentFree**
Initial version to approve, based on the version from realm-dotnet examples dir plus a couple of small fixes to improve the iOS login GUI and fix the initial loading bug where a very detailed drawing would appear to partly sync.

### CODE REVIEW

File: Xamarin/DrawXShared/DrawXSettings.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+using Realms;
+
+namespace DrawXShared
+{
+    // just stash one instance
```
Comment (kristiandupont) [inline Xamarin/DrawXShared/DrawXSettings.cs]: This comment seems a bit strange.

File: Xamarin/DrawXShared/DrawXSettingsManager.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+using System;
+using System.Diagnostics;
+using System.Linq;
+using Realms;
+
+namespace DrawXShared
+{
+    internal static class DrawXSettingsManager
+    {
+        private static Realm _realmLocalSettings;
```
Comment (kristiandupont) [inline Xamarin/DrawXShared/DrawXSettingsManager.cs]: Shouldn't this be called `_localSettingsRealm`?

File: Xamarin/DrawXShared/DrawXSettingsManager.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+                return _savedSettings;
+            }
+        }
+
+        internal static void InitLocalSettings()
+        {
+            var settingsConf = new RealmConfiguration("DrawXsettings.realm");
+            settingsConf.ObjectClasses = new[] { typeof(DrawXSettings) };
+            settingsConf.SchemaVersion = 2;  // set explicitly and bump as we add setting properties
```
Comment (kristiandupont) [inline Xamarin/DrawXShared/DrawXSettingsManager.cs:57]: Are we at schema version 2 before even releasing?

File: Xamarin/DrawXShared/ViewControllerShared.cs [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+using System.Diagnostics;
+using DrawXShared;
+using Foundation;
+using SkiaSharp.Views.iOS;
+using UIKit;
+
+namespace DrawX.IOS
+{
+    // Most of the ViewController logic is factored out here so we can subclass
+    // with a local copy in case we want a debug build which bypasses nuget as a Realm source
```
Comment (kristiandupont) [inline Xamarin/DrawXShared/ViewControllerShared.cs:29]: I don't think this comment belongs in example code

File: Xamarin/DrawXShared/ViewControllerShared.cs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+            var sb = UIStoryboard.FromName("LoginScreen", null);
+            var loginVC = sb.InstantiateViewController("Login") as LoginViewController;
+            loginVC.OnCloseLogin = (bool changedServer) =>
+            {
+                DismissModalViewController(false);
+                if (changedServer || _drawer == null)
+                {
+                    if (DrawXSettingsManager.HasCredentials())
+                    {
+                        SetupDrawer();  // pointless unless contact server
```
Comment (kristiandupont) [inline Xamarin/DrawXShared/ViewControllerShared.cs]: I don't understand this comment.

File: Xamarin/DrawXShared/SwatchColor.cs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+                {
+                    InitColors();
+                }
+
+                return _colors;
+            }
+        }
+
+        // iterating this list gives you the colours in our desired order
+        public static List<SwatchColor> Colors
```
Comment (kristiandupont) [inline Xamarin/DrawXShared/SwatchColor.cs:45]: Where is this used? Wouldn't `ColorsByName.Values` be sufficient?

File: Xamarin/DrawXShared/SwatchColor.cs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        private static Dictionary<string, SwatchColor> _colors;
+        private static List<SwatchColor> _indexedColors;
+
+        public static Dictionary<string, SwatchColor> ColorsByName
+        {
+            get
+            {
+                if (_colors == null)
+                {
+                    InitColors();
```
Comment (kristiandupont) [inline Xamarin/DrawXShared/SwatchColor.cs:37]: Does this need to be lazy? Seems complicated for no apparent performance gain.

File: Xamarin/DrawXShared/RealmDraw.cs [hunk 8]
```diff
@@ -0,0 +1,552 @@
...
+        private bool _redrawPathsAtNextDraw = true;
+        #endregion
+
+        #region Touch Areas
+        private SKRect _loginIconRect;
+        private SKRect _loginIconTouchRect;
+        //// setup in DrawBackground
+        private float _pencilWidth;
+        private float _pencilsTop;
+        private int _numPencils;
```
Comment (kristiandupont) [inline Xamarin/DrawXShared/RealmDraw.cs:111]: Do we need a field for this?

File: Xamarin/DrawXShared/RealmDraw.cs [hunk 9]
```diff
@@ -0,0 +1,552 @@
...
+
+        private SwatchColor _currentColor;
+
+        private SwatchColor CurrentColor
+        {
+            get
+            {
+                if (string.IsNullOrEmpty(_currentColor.Name))
+                {
+                    InitCurrentColor();
```
Comment (kristiandupont) [inline Xamarin/DrawXShared/RealmDraw.cs:131]: Couldn't there just be a default initial color? Also, why do we need to state variables for the active color?

File: Xamarin/DrawXShared/RealmDraw.cs [hunk 10]
```diff
@@ -0,0 +1,552 @@
...
+        }
+
+        internal void InvalidateCachedPaths()
+        {
+            _redrawPathsAtNextDraw = true;
+            _hasSavedBitmap = false;
+            _currentlyDrawing = null;
+        }
+
+        internal async void LoginToServerAsync()
```
Comment (kristiandupont) [inline Xamarin/DrawXShared/RealmDraw.cs:179]: Hm, this method is really complex. I don't have time to look into it properly right now but it seems that an example like this is going to look scary where we want it to look simple.. :-(

File: Xamarin/DrawXShared/RealmDraw.cs [hunk 11]
```diff
@@ -0,0 +1,552 @@
...
+            var s = Settings;
+            //// TODO allow entering Create User flag on credentials to pass in here instead of false
+            var credentials = Credentials.UsernamePassword(s.Username, s.Password, false);
+            var user = await User.LoginAsync(credentials, new Uri($"http://{s.ServerIP}"));
+            Debug.WriteLine($"Got user logged in with refresh token {user.RefreshToken}");
+
+            var loginConf = new SyncConfiguration(user, new Uri($"realm://{s.ServerIP}/~/Draw"));
+            _realm = Realm.GetInstance(loginConf);
+            _allPaths = _realm.All<DrawPath>() as IQueryable<DrawPath>;
+            _allPaths.SubscribeForNotifications((sender, changes, error) =>
```
Comment (kristiandupont) [inline Xamarin/DrawXShared/RealmDraw.cs:200]: This lambda should probably be factored out as it doesn't seem related to logging in..

File: Xamarin/DrawXShared/DrawXSettingsManager.cs (hunk 3 above)
Comment (AndyDentFree) [reply Xamarin/DrawXShared/DrawXSettingsManager.cs:57]: historically, yes there was a schema update as the Draw evolved

File: Xamarin/DrawXShared/SwatchColor.cs (hunk 6 above)
Comment (AndyDentFree) [reply Xamarin/DrawXShared/SwatchColor.cs:45]: As the comment says, it provides an order to the colours. Just using `ColorsByName.Values` fails to guarantee order.

File: Xamarin/DrawXShared/RealmDraw.cs (hunk 8 above)
Comment (AndyDentFree) [reply Xamarin/DrawXShared/RealmDraw.cs:111]: I guess it could be replaced by using the static `SwatchColor.ColorsByName.Count` a couple of times if you would rather not cache it.

File: Xamarin/DrawXShared/RealmDraw.cs (hunk 9 above)
Comment (AndyDentFree) [reply Xamarin/DrawXShared/RealmDraw.cs:131]: It remembers the last colour used in settings (by name) which is behaviour in the original Draw. 

Not sure I understand _Also, why do we need to state variables for the active color?_  or are you suggesting we just use `SwatchColor.Colors.ElementAt(_currentColorIndex)` and avoid having the property `CurrentColor`?

File: Xamarin/DrawXShared/RealmDraw.cs (hunk 11 above)
Comment (AndyDentFree) [reply Xamarin/DrawXShared/RealmDraw.cs:200]: Would factoring out the lambda provide enough simplification of `LoginToServerAsync`?

File: Xamarin/DrawXShared/ViewControllerShared.cs (hunk 4 above)
Comment (AndyDentFree) [reply Xamarin/DrawXShared/ViewControllerShared.cs:29]: I felt a need to explain why the VC was factored into two classes. I don't want to combine them in case doing debugging with a local variant again.

**Review by kristiandupont — APPROVED** [review]
